src/metrics/include/metrics/v2/event.hpp

```diff
@@ -0,0 +1,279 @@
+// Copyright 2019-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#ifndef HEADER_960919A5_5455_4DD2_BC68_EFBAEB228BB0_INCLUDED
+#define HEADER_960919A5_5455_4DD2_BC68_EFBAEB228BB0_INCLUDED
+
+#include <atomic>
+#include <cstdlib>
+
+#include <boost/log/trivial.hpp>
+
+#include <grpcpp/client_context.h>
+#include <grpcpp/create_channel.h>
+#include <grpcpp/security/credentials.h>
+
+#include <metrics/operation.hpp>
+#include <poplarlib/collector.grpc.pb.h>
+
+/**
+ * @namespace genny::metrics::internals::v2 this namespace is private and only intended to be used
+ * by genny's own internals. No types from the genny::metrics::v2 namespace should ever be typed
+ * directly into the implementation of an actor.
+ */
+
+namespace genny::metrics {
+
+template <typename Clocksource>
+class OperationEventT;
+
+namespace internals::v2 {
+
+class PoplarRequestError : public std::runtime_error {
+public:
+    using std::runtime_error::runtime_error;
+};
+
+
+/**
+ * Manages the collector stub and its interactions.
+ */
+class CollectorStubInterface {
+public:
+    CollectorStubInterface() {
+        if (!_stub) {
+            auto channel =
+                grpc::CreateChannel("localhost:2288", grpc::InsecureChannelCredentials());
+            _stub = poplar::PoplarEventCollector::NewStub(channel);
+        }
+    }
+
+    poplar::PoplarEventCollector::StubInterface* operator->() {
+        return _stub.get();
+    }
+
+private:
+    // We should only have one channel across all threads.
+    static std::unique_ptr<poplar::PoplarEventCollector::StubInterface> _stub;
+};
+
+
+/**
+ * Manages the stream of poplar EventMetrics.
+ */
+class StreamInterface {
+public:
+    StreamInterface(CollectorStubInterface& stub)
+        : _options{}, _response{}, _context{}, _stream{stub->StreamEvents(&_context, &_response)} {
+        _options.set_no_compression().set_buffer_hint();
+    }
+
+    void write(const poplar::EventMetrics& event) {
```

Comment:
After about 30 minutes, writes start failing because the grpc server kicks us out due to too many pings. I think a thread-safe global channel refresh might be necessary, unless we can figure out channel options to prevent that. The default C++ channel options are compatible between client and server, but the language-agnostic documentation doesn't say if that's necessarily the case between the Go server defaults and the C++ client defaults.

Reply:
What do you mean by "too many pings" and "global channel refresh"? I had a hard time finding good docs on the client stuff so if you've found a good source I wanna catch up.

Reply:
I haven't found that many details, I just noticed that after 30 minutes grpc gives me a a`GOAWAY` with error code `ENHANCE_YOUR_CALM` due to `too_many_pings`. Apparently that can happen depending on the ping settings of the client and the settings of the server ([go](https://godoc.org/google.golang.org/grpc/keepalive) doc, [C++](https://github.com/grpc/grpc/blob/master/doc/keepalive.md) doc). Might ping cedar people tomorrow to see if there are any weird settings for server enforcement.

By "global channel refresh" I mean taking the class-static channel variable, acquiring a lock, and replacing it with a new one. (I recognize that we don't want those explicit locks, but if it's brief and every 30 minutes maybe it's not so bad?) That's a heavy-handed approach though and I'd rather not need to re-connect. There might be another option in the API but I'm still learning there.

Only detailed references I can find beyond the gRPC tutorials are the [API reference](https://grpc.github.io/grpc/cpp/index.html) and the source repo.

Reply:
This was resolved by disabling the BDP probe.